aten/src/ATen/functorch/BatchRulesScatterOps.cpp

```diff
@@ -1056,6 +1056,132 @@ std::tuple<Tensor,optional<int64_t>> masked_fill_scalar_batch_rule(
+
+    auto batch_size = get_bdim_size2(self, self_bdim, index, index_bdim);
+    self_ = ensure_has_bdim(self_, self_bdim.has_value(), batch_size);
+    index_ = ensure_has_bdim(index_, index_bdim.has_value(), batch_size);
+
+    if (!inplace){
+      if (!self_bdim.has_value())
+        self_ = self_.clone();
+    }
+
+    for (const auto i : c10::irange(0, batch_size)) {
+      const auto& self_slice = self_.select(0, i);
+      const auto& index_slice = index_.select(0, i);
+      self_slice.index_fill_(
+        dim,
+        index_slice,
+        value
+      );
+    }
+    return std::make_tuple(self_, 0);
+}
+
+std::tuple<Tensor,optional<int64_t>> index_fill_int_tensor_batch_rule_impl(
+    Tensor & self, optional<int64_t> self_bdim,
+    int dim,
+    const Tensor & index, optional<int64_t> index_bdim,
+    const Tensor & value, optional<int64_t> value_bdim,
+    const bool inplace) {
+
+    const auto self_logical_rank = rankWithoutBatchDim(self, self_bdim);
+    Tensor self_ = moveBatchDimToFront(self, self_bdim);
+    Tensor index_ = moveBatchDimToFront(index, index_bdim);
+    Tensor value_ = moveBatchDimToFront(value, value_bdim);
+    dim = maybe_wrap_dim(dim, self_logical_rank);
+
+    if (inplace && !self_bdim.has_value()) {
+      vmapIncompatibleInplaceError("index_fill_");
+    }
+
+    if (!index_bdim && !value_bdim) {
+      self_.unsqueeze_(-1);
+      self_.index_fill_(dim + 1, index_, value);
+      self_.squeeze_(-1);
+      return std::make_tuple(self_, 0);
+    }
+
+    auto batch_size = get_bdim_size3(self, self_bdim, index, index_bdim, value, value_bdim);
+    self_ = ensure_has_bdim(self_, self_bdim.has_value(), batch_size);
+    index_ = ensure_has_bdim(index_, index_bdim.has_value(), batch_size);
+    value_ = ensure_has_bdim(value_, value_bdim.has_value(), batch_size);
+
+    if (!inplace){
+      if (!self_bdim.has_value())
+        self_ = self_.clone();
+    }
+
+    for (const auto i : c10::irange(0, batch_size)) {
+      const auto& self_slice = self_.select(0, i);
+      const auto& index_slice = index_.select(0, i);
+      const auto& value_slice = value_.select(0, i);
+      self_slice.index_fill_(
+        dim,
+        index_slice,
+        value_slice
+      );
+    }
+    return std::make_tuple(self_, 0);
+}
+
+std::tuple<Tensor,optional<int64_t>> index_fill__int_scalar_batch_rule(
+    Tensor & self, optional<int64_t> self_bdim,
+    int dim,
+    const Tensor & index, optional<int64_t> index_bdim,
+    const Scalar & value) {
+    return index_fill_int_scalar_batch_rule_impl(self, self_bdim, dim, index, index_bdim, value, true);
+}
+
+std::tuple<Tensor,optional<int64_t>> index_fill__int_tensor_batch_rule(
+    Tensor & self, optional<int64_t> self_bdim,
+    int dim,
+    const Tensor & index, optional<int64_t> index_bdim,
+    const Tensor & value, optional<int64_t> value_bdim) {
+    return index_fill_int_tensor_batch_rule_impl(self, self_bdim, dim, index, index_bdim, value, value_bdim, true);
+}
```

Comment:
The signature for inplace batching rules is actually

```
void index_fill__int_scalar_batch_rule(...) {
  ...
}
```
Since the operation is in-place, we just return `self` and `self_bdim`. This happens somewhere in the codegen for vmap (I can link it if you're interested). The codegen ends up ignoring the return value of this function. To make it clearer, we should change the signature to return void.

Reply:
This makes sense. I'm indeed interested in the codegen part, can you please point out the link?

Reply:
I should probably toss this into a guide somewhere, but:
- here is the [codegen](https://github.com/pytorch/pytorch/blob/67c53d50e5b34a536cce2ad98b9ca5f65ff8a34d/torchgen/gen_vmap_plumbing.py#L166-L176)
- it generates a `{operator}_generated_plumbing` for each PyTorch ATen operator. You can see the output in the local build/aten/src/ATen/VmapGeneratedPlumbing.h file after you build PyTorch. [example](https://gist.github.com/zou3519/2e5d020d04b192c714121993da6daeeb)
- the VMAP_SUPPORT(operator, batch_rule) macro is just `{operator}_generated_plumbing<decltype(batch_rule), batch_rule>`.
- you'll notice that the plumbing doesn't actually use the return value of `batch_rule`, so it can be whatever. To avoid confusion, we should return nothing from index_fill__int_tensor_batch_rule

Reply:
Thanks for the info! So based on what I found, since the in-place plumbing doesn't take any return value from the batch_rule, the batch dims that seem to me are expected to be no change after the batch_rule. As a result, should we move the batch dims back before returning from the batch_rule?

From the implementation of `_index_put_impl__batch_rule`, for example, it seems not moving the batch dims back.

Reply:
When we move the batch dims to the front, usually we use at::movedim (or equivalent). This produces a new Tensor that is a view of the original one. In this case, there's no need to move the batch dims back -- the original tensor still has the batch dim in the correct position.